### PR TITLE
support collation in create query

### LINKED ISSUE #35892: Parse collations in CREATE TABLE, throw exception or ignore.
**Use case**

Easy migration from MySQL.

**Describe the solution you'd like**

Syntax like `x varchar(255) COLLATE utf8_unicode_ci NOT NULL,` should be parsed and supported on AST level.
Interpreter should throw exception if collation is provided, unless a setting `compatibility_ignore_collation_in_create_table` is set to 1 (in that case they are ignored).

### PR DESCRIPTION
**Pull request by yuuch**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Parse collations in CREATE TABLE, throw exception or ignore. closes #35892

### CODE REVIEW

**Comment by yuuch** [comment]
@Mergifyio update

**Comment by yuuch** [comment]
@vdimir @alexey-milovidov Can someone review this PR?

**Comment by alexey-milovidov** [comment]
Someone from my team will take this PR for review.

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 1]
```diff
@@ -460,8 +460,12 @@ ColumnsDescription InterpreterCreateQuery::getColumnsDescription(
     {
         const auto & col_decl = ast->as<ASTColumnDeclaration &>();
 
-        DataTypePtr column_type = nullptr;
+        if (col_decl.collation && !context_->getSettingsRef().compatibility_ignore_collation_in_create_table)
+        {
+            throw Exception("Cannot support collation, please set compatibility_ignore_collation_in_create_table=true", ErrorCodes::LOGICAL_ERROR);
```
Comment (vdimir) [inline src/Interpreters/InterpreterCreateQuery.cpp:465]:
```suggestion
            throw Exception("Cannot support collation, please set compatibility_ignore_collation_in_create_table=true", ErrorCodes::NOT_IMPLEMENTED);
```

File: src/Parsers/ASTCollation.cpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#include <Parsers/ASTCollation.h>
+
+namespace DB
+{
+    ASTPtr ASTCollation::clone() const
+    {
+        auto res = std::make_shared<ASTCollation>(*this);
+        res->collation = collation;
```
Comment (vdimir) [inline src/Parsers/ASTCollation.cpp:8]: Shouldn't we clone recursively?
```suggestion
        res->collation = collation->clone();
```

File: tests/queries/0_stateless/02269_create_table_with_collation.sql [hunk 3]
```diff
@@ -0,0 +1,7 @@
+SET compatibility_ignore_collation_in_create_table=false;
+CREATE TABLE t_02267_collation (x varchar(255) COLLATE utf8_unicode_ci NOT NULL) ENGINE = Memory; --{serverError 49}
+
+SET compatibility_ignore_collation_in_create_table=true;
+CREATE TABLE t_02267_collation (x varchar(255) COLLATE utf8_unicode_ci NOT NULL) ENGINE = Memory;
+
```
Comment (vdimir) [inline tests/queries/0_stateless/02269_create_table_with_collation.sql:6]: It's also good to add tests to check that it properly formatted with `EXPLAIN SYNTAX`

File: tests/queries/0_stateless/02269_create_table_with_collation.sql [hunk 4]
```diff
@@ -0,0 +1,7 @@
+SET compatibility_ignore_collation_in_create_table=false;
+CREATE TABLE t_02267_collation (x varchar(255) COLLATE utf8_unicode_ci NOT NULL) ENGINE = Memory; --{serverError 49}
```
Comment (vdimir) [inline tests/queries/0_stateless/02269_create_table_with_collation.sql:2]:
```suggestion
CREATE TABLE t_02267_collation (x varchar(255) COLLATE utf8_unicode_ci NOT NULL) ENGINE = Memory; -- {serverError NOT_IMPLEMENTED}
```

File: src/Parsers/ParserCreateQuery.h [hunk 5]
```diff
@@ -183,6 +186,11 @@ bool IParserColumnDeclaration<NameParser>::parseImpl(Pos & pos, ASTPtr & node, E
             return false;
         if (!type_parser.parse(pos, type, expected))
             return false;
+        if (s_collate.ignore(pos, expected))
+        {
+            if (!collation_parser.parse(pos, collation_expression, expected))
```
Comment (vdimir) [inline src/Parsers/ParserCreateQuery.h:191]: How we should handle case when there's no keyword between `COLLATE` and `NOT NULL` ?

```
:) EXPLAIN SYNTAX CREATE TABLE t (x varchar(255) COLLATE NOT NULL) ENGINE = Memory;

EXPLAIN SYNTAX
CREATE TABLE t
(
    `x` varchar(255) NULL COLLATION NOT
)
ENGINE = Memory
```


```
:) EXPLAIN SYNTAX CREATE TABLE t (x varchar(255) COLLATE NULL) ENGINE = Memory;

EXPLAIN SYNTAX
CREATE TABLE t
(
    `x` varchar(255) COLLATION `NULL`
)
ENGINE = Memory
```

File: src/Parsers/ASTColumnDeclaration.cpp [hunk 6]
```diff
@@ -97,6 +102,12 @@ void ASTColumnDeclaration::formatImpl(const FormatSettings & settings, FormatSta
         settings.ostr << ' ' << (settings.hilite ? hilite_keyword : "") << "TTL" << (settings.hilite ? hilite_none : "") << ' ';
         ttl->formatImpl(settings, state, frame);
     }
+
+    if (collation)
+    {
+        settings.ostr << ' ' << (settings.hilite ? hilite_keyword : "") << "COLLATION" << (settings.hilite ? hilite_none : "") << ' ';
```
Comment (vdimir) [inline src/Parsers/ASTColumnDeclaration.cpp:108]: We parse `COLLATE` but format as `COLLATION`
```suggestion
        settings.ostr << ' ' << (settings.hilite ? hilite_keyword : "") << "COLLATE" << (settings.hilite ? hilite_none : "") << ' ';
```

**Review by vdimir — COMMENTED** [review]
Sorry for long response. 
Generally fine, just some comments about formatting and parsing corner cases like (`COLLATE NULL`)

File: src/Parsers/ParserCreateQuery.h (hunk 5 above)
Comment (yuuch) [reply src/Parsers/ParserCreateQuery.h:191]: I will check  something after the COLLATE  to make sure it is the name of validate collation (e.g.`utf8mb4_0900_ai_ci`, `latin1_german1_ci`...etc, total 273 collations got from mysql's `INFORMATION_SCHEMA.COLLATION`

File: tests/queries/0_stateless/02269_create_table_with_collation.sql (hunk 3 above)
Comment (yuuch) [reply tests/queries/0_stateless/02269_create_table_with_collation.sql:11]: Ok , I will add `EXPLAIN SYNTAX` to the test file.

File: src/Parsers/ExpressionElementParsers.cpp [hunk 7]
```diff
@@ -1309,6 +1310,309 @@ bool ParserCodec::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
     return true;
 }
 
+const char * ParserCollation::valid_collations[] = {
```
Comment (vdimir) [inline src/Parsers/ExpressionElementParsers.cpp:1313]: It's a bit overkill, need just expect `isWordCharASCII`

Comment (yuuch) [reply src/Parsers/ExpressionElementParsers.cpp:1313]: > It's a bit overkill, need just expect `isWordCharASCII`

I feel confused. If I try to parse `... COLLATE NOT NULL`, will  `isWordCharASCII` regard the `NOT` as `false` ?

Comment (vdimir) [reply src/Parsers/ExpressionElementParsers.cpp]: Good point! In that case maybe just check suffix https://github.com/ClickHouse/ClickHouse/pull/36271/commits/f6f58f7609600f2ccc8479e015ac00a794894846

Comment (yuuch) [reply src/Parsers/ExpressionElementParsers.cpp]: > Good point! In that case maybe just check suffix [f6f58f7](https://github.com/ClickHouse/ClickHouse/commit/f6f58f7609600f2ccc8479e015ac00a794894846)

Good Job

File: src/Parsers/ExpressionElementParsers.cpp [hunk 8]
```diff
@@ -9,6 +9,7 @@
 #include <Common/typeid_cast.h>
```
Comment (vdimir) [inline src/Parsers/ExpressionElementParsers.cpp:9]:
```suggestion
#include <Common/typeid_cast.h>
#include <Common/StringUtils/StringUtils.h>
```

**Comment by vdimir** [comment]
> ClickHouse special build check (actions) — 7/8 artifact groups are OK

:thinking:
```
/build_docker/contrib/bzip2-cmake -nodefaultlibs /usr/lib/llvm-13/lib/clang/13.0.1/lib/linux/libclang_rt.builtins-x86_64.a  -lc -lm -lrt -lpthread -ldl && :
May 03 13:05:02 ld.lld-13: error: undefined symbol: detail::endsWith(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const*, unsigned long)
May 03 13:05:02 >>> referenced by StringUtils.h:37 (../src/Common/StringUtils/StringUtils.h:37)
May 03 13:05:02 >>>               src/Parsers/CMakeFiles/clickhouse_parsers.dir/ExpressionElementParsers.cpp.o:(DB::ParserCollation::parseImpl(DB::IParser::Pos&, std::__1::shared_ptr<DB::IAST>&, DB::Expected&))
May 03 13:05:02 >>> referenced by StringUtils.h:37 (../src/Common/StringUtils/StringUtils.h:37)
May 03 13:05:02 >>>               src/Parsers/CMakeFiles/clickhouse_parsers.dir/ExpressionElementParsers.cpp.o:(DB::ParserCollation::parseImpl(DB::IParser::Pos&, std::__1::shared_ptr<DB::IAST>&, DB::Expected&))
May 03 13:05:02 >>> referenced by StringUtils.h:37 (../src/Common/StringUtils/StringUtils.h:37)
May 03 13:05:02 >>>               src/Parsers/CMakeFiles/clickhouse_parsers.dir/ExpressionElementParsers.cpp.o:(DB::ParserCollation::parseImpl(DB::IParser::Pos&, std::__1::shared_ptr<DB::IAST>&, DB::Expected&))
May 03 13:05:02 >>> referenced 1 more times
May 03 13:05:02 clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Review by vdimir — APPROVED** [review]

**Pull request merged**